Client/Frontend/Widgets/BottomSheetViewController.swift

```diff
@@ -0,0 +1,248 @@
+        super.viewDidAppear(animated)
+        show()
+    }
+    
+    override func viewDidLayoutSubviews() {
+        yPosition = contentView.isHidden ? view.frame.maxY : initialDrawerYPosition
+        
+        updateDrawerViewConstraints()
+    }
+    
+    override func traitCollectionDidChange(_ previousTraitCollection: UITraitCollection?) {
+        super.traitCollectionDidChange(previousTraitCollection)
+        updateDrawerViewConstraints()
+    }
+    
+    // MARK: - Constraints setup
+    
+    private func makeConstraints() {
+        backgroundOverlayView.snp.makeConstraints { make in
+            let bottomInset = parent?.view.safeAreaInsets.bottom ?? 0
+            make.leading.trailing.top.equalToSuperview()
+            make.bottom.equalToSuperview().offset(bottomInset)
+        }
+        
+        handleView.snp.remakeConstraints {
+            $0.width.equalTo(UX.HandleWidth)
+            $0.height.equalTo(UX.HandleHeight)
+
+            $0.centerX.equalTo(contentView)
+            $0.top.equalTo(contentView).offset((UX.HandleMargin - UX.HandleHeight) / 2)
+        }
+        
+        closeButton.snp.makeConstraints {
+            $0.top.equalToSuperview().inset(10)
+            $0.right.equalToSuperview().inset(7)
+            $0.size.equalTo(26)
+        }
+        
+        updateDrawerViewConstraints()
+    }
+    
+    private func updateDrawerViewConstraints() {
+        let allCorners: CACornerMask = [.layerMinXMinYCorner, .layerMaxXMinYCorner,
+                          .layerMinXMaxYCorner, .layerMaxXMaxYCorner]
+        
+        let onlyTopCorners: CACornerMask = [.layerMinXMinYCorner, .layerMaxXMinYCorner]
+        contentView.layer.maskedCorners = showAsPopup ? allCorners : onlyTopCorners
+        
+        handleView.isHidden = showAsPopup
+        
+        // Don't remake constraints if not needed.
+        if yPosition == initialDrawerYPosition { return }
+        
+        contentView.snp.remakeConstraints {
+            if showAsPopup || UIApplication.shared.statusBarOrientation.isLandscape {
+                $0.bottom.equalToSuperview()
+                $0.centerX.equalToSuperview()
+                $0.width.equalTo(maxHorizontalWidth)
+            } else {
+                $0.leading.trailing.equalToSuperview()
+                $0.bottom.equalToSuperview()
+            }
+        }
+    }
+    
+    // MARK: - Animations
+    
+    @objc private func didRecognizePanGesture(_ gestureRecognizer: UIPanGestureRecognizer) {
+        // The view shouldn't be draggable on iPads
+        if showAsPopup { return }
+
+        let translation = gestureRecognizer.translation(in: contentView)
+        yPosition += translation.y
+        gestureRecognizer.setTranslation(CGPoint.zero, in: contentView)
+
+        if gestureRecognizer.state != .ended { return }
+
+        let velocity = gestureRecognizer.velocity(in: contentView).y
+        let landingYPosition = yPosition + velocity / 10
```

Comment:
Use projected end point algorithm:

```swift
// Distance travelled after decelerating to zero velocity at a constant rate (credit: a WWDC video)
func project(initialVelocity: Float, decelerationRate: Float) -> Float {
  return (initialVelocity / 1000.0) * decelerationRate / (1.0 - decelerationRate) 
}
```
Where deceleration rate can be taken from UIScrollView.DecelerationRate.normal

You can see its usage already here: https://github.com/brave/brave-ios/blob/24c14e71ece89cab32a168d7b62d927eeed29291/BraveRewardsUI/Ads/AdsViewController.swift#L174-L208

Reply:
thank you for roasting this class, all this logic was taken from firefox and then adjusted by me